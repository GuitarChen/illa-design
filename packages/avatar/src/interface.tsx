import { CSSProperties, HTMLAttributes, ReactNode } from "react"

export type AvatarColorScheme =
  | string
  | "white"
  | "blackAlpha"
  | "gray"
  | "grayBlue"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "cyan"
  | "purple"
  | "techPink"
  | "techPurple"

export type AvatarShape = "circle" | "square"

export type AvatarSize = "small" | "medium" | "large"

export interface AvatarProps extends HTMLAttributes<HTMLDivElement> {
  colorScheme?: AvatarColorScheme
  size?: AvatarSize
  shape?: AvatarShape
  text?: string
  src?: string
  icon?: ReactNode
}

export interface AvatarGroupProps
  extends HTMLAttributes<HTMLDivElement>,
    AvatarGroupContextProps {}

export interface AvatarGroupContextProps {
  zIndexAscend?: boolean
  maxCount?: number
  colorScheme?: AvatarColorScheme
  size?: AvatarSize
  style?: CSSProperties
}
