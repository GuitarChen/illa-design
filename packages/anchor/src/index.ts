export * from "./anchor"
export * from "./interface"
