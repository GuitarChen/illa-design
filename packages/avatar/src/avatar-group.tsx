import { Children, CSSProperties, forwardRef, ReactNode } from "react"
import { AvatarGroupProps } from "./interface"
import { css } from "@emotion/react"
import { Avatar } from "./avatar"
import { globalColor, illaPrefix } from "@illa-design/theme"
import { AvatarGroupContext } from "./avatar-group-context"

const avatarGroupCss = css`
  display: flex;
  flex-direction: row;
  align-items: center;
`

export const AvatarGroup = forwardRef<HTMLDivElement, AvatarGroupProps>(
  (props, ref) => {
    const {
      zIndexAscend,
      maxCount = Number.MAX_VALUE,
      colorScheme,
      size,
      ...otherProps
    } = props

    const childrenCount = Children.count(props.children)
    let nodeList: ReactNode[]
    if (childrenCount <= maxCount) {
      nodeList = Children.toArray(props.children)
    } else {
      nodeList = Children.toArray(props.children).splice(0, maxCount)
    }
    if (childrenCount > maxCount) {
      nodeList = [
        ...nodeList,
        <Avatar colorScheme={"gray"} text={`+${childrenCount - maxCount}`} />,
      ]
    }

    let marginLeft: string
    switch (props.size) {
      default:
      case "small":
        marginLeft = "-8px"
        break
      case "medium":
        marginLeft = "-12px"
        break
      case "large":
        marginLeft = "-20px"
        break
    }
    const newNodeList = Children.map(nodeList, (child, index) => {
      const isFirstAvatar = index === 0

      const style = {
        marginLeft: `${isFirstAvatar ? "0px" : marginLeft}`,
        border: `solid 2px ${globalColor(`--${illaPrefix}-white-01`)}`,
        borderRadius: "50%",
        zIndex: `${zIndexAscend ? index : nodeList.length - index}`,
      } as CSSProperties

      return (
        <AvatarGroupContext.Provider
          value={{
            zIndexAscend,
            maxCount,
            colorScheme,
            size,
            style,
          }}
        >
          {child}
        </AvatarGroupContext.Provider>
      )
    })
    return (
      <div
        css={css`
          display: inline-block;
          ${avatarGroupCss};
        `}
        ref={ref}
        {...otherProps}
      >
        {newNodeList}
      </div>
    )
  },
)

AvatarGroup.displayName = "AvatarGroup"
