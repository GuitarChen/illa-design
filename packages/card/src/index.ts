export * from "./card"
export * from "./interface"
export * from "./card-grid"
export * from "./meta"
