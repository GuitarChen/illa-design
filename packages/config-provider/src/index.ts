export * from "./config-provider"
export * from "./interface"
export * from "./config-provider-context"
export * from "./config"

// locale
export * from "./locale/en-US"
export * from "./locale/zh-CN"
export * from "./locale/def"
