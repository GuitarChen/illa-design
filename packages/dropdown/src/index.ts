export * from "./dropdown"
export * from "./interface"
