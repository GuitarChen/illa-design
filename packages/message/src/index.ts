export * from "./style"
export * from "./message"
export * from "./interface"
