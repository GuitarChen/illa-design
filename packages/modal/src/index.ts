export * from "./modal"
export * from "./interface"
export * from "./style"
export * from "./portal"
export * from "./config"
export * from "./confirm"
export * from "./useModal/hookModal"
export * from "./useModal/contextHolder"
export * from "./useModal/index"
