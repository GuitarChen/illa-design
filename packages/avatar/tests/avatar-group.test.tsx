import { Avatar, AvatarGroup } from "../src"
import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"

test("AvatarGroup renders correctly", () => {
  render(
    <AvatarGroup data-testid="test-avatar-group">
      <Avatar text="A" />
      <Avatar text="B" />
      <Avatar text="C" />
    </AvatarGroup>,
  )
  expect(screen.getByTestId("test-avatar-group")).toBeInTheDocument()
})

test("AvatarGroup renders with max count", () => {
  render(
    <AvatarGroup data-testid="test-with-mac-count" maxCount={3}>
      <Avatar text="A" />
      <Avatar text="B" />
      <Avatar text="C" />
      <Avatar text="D" />
      <Avatar text="E" />
    </AvatarGroup>,
  )
  expect(screen.getByTestId("test-with-mac-count").childElementCount).toBe(4)
  expect(screen.getByText("+2")).toBeInTheDocument()
})

test("AvatarGroup renders with zIndexAscend", () => {
  const { asFragment } = render(
    <AvatarGroup zIndexAscend={true}>
      <Avatar text="A" />
      <Avatar text="B" />
      <Avatar text="C" />
      <Avatar text="D" />
      <Avatar text="E" />
    </AvatarGroup>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("AvatarGroup renders with large props", () => {
  const { asFragment } = render(
    <AvatarGroup size="large" colorScheme="yellow">
      <Avatar text="A" />
      <Avatar text="B" />
      <Avatar text="C" />
      <Avatar text="D" />
      <Avatar text="E" />
    </AvatarGroup>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("AvatarGroup renders with medium props", () => {
  const { asFragment } = render(
    <AvatarGroup size="medium" colorScheme="yellow">
      <Avatar text="A" />
      <Avatar text="B" />
      <Avatar text="C" />
      <Avatar text="D" />
      <Avatar text="E" />
    </AvatarGroup>,
  )
  expect(asFragment()).toMatchSnapshot()
})
