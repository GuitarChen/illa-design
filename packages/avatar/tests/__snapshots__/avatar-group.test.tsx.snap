// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup renders with large props 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  vertical-align: middle;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-2 {
  background-color: #e07800;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 64px;
  height: 64px;
  font-size: 24px;
  line-height: 32px;
  border-radius: 50%;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
      style="margin-left: 0px; border: 2px solid #fff; border-radius: 50%; z-index: 5;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          A
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -20px; border: 2px solid #fff; border-radius: 50%; z-index: 4;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          B
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -20px; border: 2px solid #fff; border-radius: 50%; z-index: 3;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          C
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -20px; border: 2px solid #fff; border-radius: 50%; z-index: 2;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          D
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -20px; border: 2px solid #fff; border-radius: 50%; z-index: 1;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          E
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AvatarGroup renders with medium props 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  vertical-align: middle;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-2 {
  background-color: #e07800;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 40px;
  height: 40px;
  font-size: 14px;
  line-height: 22px;
  border-radius: 50%;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
      style="margin-left: 0px; border: 2px solid #fff; border-radius: 50%; z-index: 5;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          A
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -12px; border: 2px solid #fff; border-radius: 50%; z-index: 4;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          B
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -12px; border: 2px solid #fff; border-radius: 50%; z-index: 3;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          C
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -12px; border: 2px solid #fff; border-radius: 50%; z-index: 2;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          D
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -12px; border: 2px solid #fff; border-radius: 50%; z-index: 1;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          E
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AvatarGroup renders with zIndexAscend 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  vertical-align: middle;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-2 {
  background-color: #c2c2c2;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 32px;
  height: 32px;
  font-size: 14px;
  line-height: 22px;
  border-radius: 50%;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
      style="margin-left: 0px; border: 2px solid #fff; border-radius: 50%; z-index: 0;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          A
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -8px; border: 2px solid #fff; border-radius: 50%; z-index: 1;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          B
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -8px; border: 2px solid #fff; border-radius: 50%; z-index: 2;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          C
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -8px; border: 2px solid #fff; border-radius: 50%; z-index: 3;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          D
        </span>
      </div>
    </div>
    <div
      class="emotion-1"
      style="margin-left: -8px; border: 2px solid #fff; border-radius: 50%; z-index: 4;"
    >
      <div
        class="emotion-2"
      >
        <span
          style="transform: scale(1);"
        >
          E
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
